### ISSUE TITLE
Resolve leaves import resolutions in an inconsistent state

### ISSUE THREAD

**Issue by jseyfried** [rust-lang/rust#31403]
For example,
```rust
mod foo {
    pub mod bar {} // This defines bar in the type namespace
    pub use alpha::bar; // This defines bar in the value namespace

    // This should define baz in both namespaces, but it only defines `baz` in the type namespace.
    pub use self::bar as baz;
    pub fn baz() {} // This should collide with the imported `baz`, but now it does not.
}

pub fn f() {}
mod alpha {
    pub use self::f as bar; // Changing this to `pub fn bar() {}` makes the imports consistent
    pub use super::*;
}

fn main() {
    foo::bar();
    foo::baz(); // This errors
}
```

**Comment by jseyfried** [comment]
This is caused by [this code](https://github.com/rust-lang/rust/blob/34af2de4096b3b1c5d3a5b70171c6e27822aaefb/src/librustc_resolve/resolve_imports.rs#L423-L436), which incorrectly reports that the import of `bar` as `baz` fails in the type namespace.
